src/c128time.c

```diff
@@ -7,33 +7,83 @@
 #include <peekpoke.h>
 #include <stdlib.h>
 #include <stdio.h>
-#include "general.h"
+#include <c128.h>
 #include <stdint.h>
 
-char *get_time(void) {
-  uint32_t h = PEEK(0x00A0) * 65536, m = PEEK(0x00A1) * 256, s = PEEK(0x00A2);
-  static char buffer[9];
-  BYTE hrs, min;
+#include "bcd2dec.h"
+#include "general.h"
+#include "globals.h"
+
+void update_time(void) {
+  uint8_t bcd_hour, hour, min, sec, tenth;
+
+  /* Read the hour register first to stop the clock from updating the external
+   * registers from the internal (still ticking!) CIA registers. */
+
+  bcd_hour = CIA1.tod_hour;
+
+  /* if high bit is set, it is pm */
+  if (bcd_hour & 0x80) {
+    hour = bcd2dec(bcd_hour ^ 0x80);
+    /* adjust for 24h clock, 12:??pm is still 12:?? */
+    if (hour != 12) {
+      hour += 12;
+    }
+  } else {
+    hour = bcd2dec(bcd_hour);
+  }
+
+  sec = bcd2dec(CIA1.tod_sec);
+  min = bcd2dec(CIA1.tod_min);
+
+  /* MUST read tod_10 to enable the clock latch again */
+  tenth = CIA1.tod_10;
```

Comment:
I suggest `(void)CIA1.tod_10;` to signal that the value is discarded immediately

Reply:
This will be optimized away. I guess the saner alternative would be inline assembler.

Reply:
This would not happen if CIA1 were declared volatile. Should we open an issue at cc65/cc65?

Reply:
Yeah, we should.